internal/secret-storage-backend/server.go

```diff
@@ -0,0 +1,315 @@
+type AdditionalParameters struct {
+	Provider string `json:"provider"`
+}
+
+var _ pb.StorageBackendServer = &Handler{}
+
+// Handler handles incoming requests to the Secret storage backend gRPC server.
+type Handler struct {
+	pb.UnimplementedStorageBackendServer
+
+	log *zap.Logger
+
+	providers map[string]tellercore.Provider
+}
+
+const (
+	lockedByField        = "locked_by"
+	firstResourceVersion = 1
+)
+
+var (
+	// NilRequestInputError describes an error with an invalid request.
+	NilRequestInputError = status.Error(codes.InvalidArgument, "request data cannot be nil")
+)
+
+// NewHandler returns new Handler.
+func NewHandler(log *zap.Logger, providers map[string]tellercore.Provider) *Handler {
+	return &Handler{
+		log:       log,
+		providers: providers,
+	}
+}
+
+// GetValue returns a value for a given TypeInstance. It returns nil as value if a given secret is not found.
+func (h *Handler) GetValue(_ context.Context, request *pb.GetValueRequest) (*pb.GetValueResponse, error) {
+	if request == nil {
+		return nil, NilRequestInputError
+	}
+
+	provider, err := h.getProviderFromAdditionalParams(request.AdditionalParameters)
+	if err != nil {
+		return nil, err
+	}
+
+	key := h.storageKeyForTypeInstanceValue(provider, request.TypeinstanceId, request.ResourceVersion)
+	entry, err := h.getEntry(provider, key)
+	if err != nil {
+		return nil, err
+	}
+
+	var value []byte
+	if entry.IsFound {
+		value = []byte(entry.Value)
+	}
+
+	return &pb.GetValueResponse{
+		Value: value,
+	}, nil
+}
+
+// GetLockedBy returns a locked by data for a given TypeInstance. It returns nil as value if a given secret is not found.
+func (h *Handler) GetLockedBy(_ context.Context, request *pb.GetLockedByRequest) (*pb.GetLockedByResponse, error) {
+	if request == nil {
+		return nil, NilRequestInputError
+	}
+
+	provider, err := h.getProviderFromAdditionalParams(request.AdditionalParameters)
+	if err != nil {
+		return nil, err
+	}
+
+	key := h.storageKeyForLockedBy(provider, request.TypeinstanceId)
+	entry, err := h.getEntry(provider, key)
+	if err != nil {
+		return nil, err
+	}
+
+	var lockedBy *string
+	if entry.IsFound && entry.Value != "" {
+		lockedBy = ptr.String(entry.Value)
+	}
+
+	return &pb.GetLockedByResponse{
+		LockedBy: lockedBy,
+	}, nil
+}
+
+// OnCreate handles TypeInstance creation by creating secret in a given provider.
+func (h *Handler) OnCreate(_ context.Context, request *pb.OnCreateRequest) (*pb.OnCreateResponse, error) {
+	if request == nil {
+		return nil, NilRequestInputError
+	}
+
+	err := h.handlePutValue(
+		request.AdditionalParameters,
+		request.TypeinstanceId,
+		firstResourceVersion,
+		request.Value,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.OnCreateResponse{}, nil
+}
+
+// OnUpdate handles TypeInstance update by updating secret in a given provider.
+func (h *Handler) OnUpdate(_ context.Context, request *pb.OnUpdateRequest) (*pb.OnUpdateResponse, error) {
```

Comment:
This is something I thought will be nice - if we move the `backend.id` to TypeInstanceResourceVersion, then there should be such case that new revision will be . But I agree, it could be problematic - I will change it now and let's revisit it here: https://github.com/capactio/capact/issues/636. Maybe we should allow changing backend id, but not for resource version, but whole TypeInstance, so, migrate it to new backend? 🤔